### DISCUSSION TITLE
Resolved: `serde` code organisation

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1886] [General]
Code that interacts with `serde` has a few properties that leads me to the idea that it might be better if it was all together in a single module instead of spread across the code base.

Notice that:
- The `serde` library is not that approachable for new Rust devs
- `serde` code often has a lot of boilerplate/logic to do with `serde` and not much to do with the implementee type (see below)
- `serde` code doesn't get touched much, for good reason serialisation should be stable and just work
- `serde_utils` has all the problems of any other use of the term "utils"
- `consensus::serde` as an abstraction is unrelated to consensus, it is a _consumer_ of the `consensus` module

I can do an exploratory PR if there is any positive response to this discussion post.

Example, this code has no logic interesting enough to require it to be next to the `U256` struct.
```rust
#[cfg(feature = "serde")]
impl crate::serde::Serialize for U256 {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: crate::serde::Serializer,
    {
        struct DisplayHex(U256);

        impl fmt::Display for DisplayHex {
            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { write!(f, "{:x}", self.0) }
        }

        if serializer.is_human_readable() {
            // TODO: fast hex encoding.
            serializer.collect_str(&DisplayHex(*self))
        } else {
            let bytes = self.to_be_bytes();
            serializer.serialize_bytes(&bytes)
        }
    }
}
```

**Comment by apoelstra** [comment]
Yeah, I like the idea of moving everything serde-related into a `serde` module. Though OTOH, because this code basically never gets touched I don't really mind having it strewn around the library. It's easy to recognize and scroll through.

OTOOH there's a lot of repetition and I find it super irritating to have all these serde macros that you need to track down when you *do* want to work with serde. If we had more macros (or preferably, trait/generic tricks) and they were more local, that'd be nicer.

**Comment by Kixunil** [comment]
I don't think it's that great. The main problems with sticking everything into a single module:

* It probably should be called `serde` but such module at the root might cause confusion - people could confuse it with `serde`, the crate. It'd also force us to keep the rename (`actual_serde`) forever.
* Crate smashing will require us to split it anyway due to orphan rules (and even then I wouldn't be happy importing whole `bitcoin` library if I only need part)
* It could end up being a big mess of unrelated things; the current hierarchy provides natural separation

Regarding your points:

>     * The `serde` library is not that approachable for new Rust devs

I don't see how this is relevant
 
>     * `serde` code often has a lot of boilerplate/logic to do with `serde` and not much to do with the implementee type (see below)

Actually this should always be the case - lack of this is probably a design problem! `serde` should always use public accessors internally to avoid breaking invariants. But even better code that protects invariants should be in its own private module to protect them from other code too.

>     * `serde` code doesn't get touched much, for good reason serialisation should be stable and just work

Correct.

>     * `serde_utils` has all the problems of any other use of the term "utils"


IIRC the name was to prevent conflict with the crate, which is currently renamed to `actual_serde`. We could change it to `serde` but note that it might be confusing. `serde_common` might be marginally better. Also this module is intentionally private which I think is a very sensible approach.

>     * `consensus::serde` as an abstraction is unrelated to consensus, it is a _consumer_ of the `consensus` module

Yes but there probably isn't a better obvious name/path for it. Putting it within `encoding` is an option but then the path becomes crazy-long.

I think putting `serde::Trait` impls **only** into a separate private module would be kinda OK but all other APIs should stay.

**Comment by tcharding** [comment]
Thanks for the in depth replies, my bad for failing to realise the naming conflict thing with `crate::serde`. So with no clear way to improve lets leave it all as is. Thanks again.

**Comment by stevenroose** [comment]
Tbh, I really appreciate impls of a type to be in the same file as the type. Like if you have Address, I appreciate the serde impl for Address to be around there and not someplace entirely different. Usually I place impls underneath the type definition, but in some of our modules the serde impls are all at the bottom, I think that's fair. They are around but out of the way.

If we're gonna have one giant serde module, that kinda breaks separation of concerns or whatever it's called, being that everything related to one concept is bundled together.
